server/appengine/src/main/java/who/RefreshCaseStatsServlet.java

```diff
@@ -100,105 +189,30 @@ protected void doGet(
     while (true) {
       String jsonString = getDashboardContents(numItems);
 
-      JsonObject root = new JsonParser().parse(jsonString).getAsJsonObject();
+      JsonObject root = JsonParser.parseString(jsonString).getAsJsonObject();
 
       JsonArray rows = root.getAsJsonArray("features");
       if (rows == null || rows.size() == 0) {
         break;
       }
       numItems += rows.size();
-      if (numItems > 100000) {
-        // We may no longer be able to process this in the 30-seconds allocated to cron requests.
-        logger.error(
-          "Ending processing - Will not update stats - More than 100000 features returned by ArcGIS endpoint."
-        );
-        // Throw a 500.
-        throw new RuntimeException(
-          "Ending processing - Will not update stats - More than 100000 features returned by ArcGIS endpoint."
-        );
-      }
-
-      // Given that each row has heterogeneous elements, not sure there is much benefit
-      // to using gson with reflection here.
-      for (JsonElement feature : rows) {
-        JsonObject featureAsJsonObject = feature.getAsJsonObject();
-        JsonElement attributesElement = featureAsJsonObject.get("attributes");
-
-        JsonObject attributes = attributesElement.getAsJsonObject();
-
-        long timestamp = attributes.get("date_epicrv").getAsLong();
-        long dailyDeaths = attributes.get("NewDeath").getAsLong();
-        long totalDeaths = attributes.get("CumDeath").getAsLong();
-        long dailyCases = attributes.get("NewCase").getAsLong();
-        long totalCases = attributes.get("CumCase").getAsLong();
-
-        updateData(
-          globalData,
-          timestamp,
-          dailyDeaths,
-          totalDeaths,
-          dailyCases,
-          totalCases
-        );
-
-        if (!(attributes.get("ISO_2_CODE") instanceof JsonNull)) {
-          String isoCode = attributes.get("ISO_2_CODE").getAsString();
-          JurisdictionData country = countryData.get(isoCode);
-          if (country == null) {
-            country = new JurisdictionData();
-            countryData.put(isoCode, country);
-          }
-          updateData(
-            country,
-            timestamp,
-            dailyDeaths,
-            totalDeaths,
-            dailyCases,
-            totalCases
-          );
-        }
-      }
+      processWhoStats(rows, countryData, globalData);
```

Comment:
I think `processWhoStats` should include the JSON parsing as well. So it takes a single string and then does all the work. Better encapsulation for testing as well.

Reply:
I think it's the opposite. processWhoStats takes in pretty structured things as input and output. the json parsing is done beforehand. note that this allows it to be done different ways (e.g. the unit test does not actually use a string).  and the outer loop right here actually uses some of the data in the parsed json which isn't really relevant to processWhoStats.